colossalai/nn/_ops/cache_embedding/parallel_freq_aware_embedding.py

```diff
@@ -0,0 +1,136 @@
+            embedding_dim, self.rank, self.world_size)
+        self.embedding_dim_per_partition = self.partition_end_index - self.partition_start_index
+
+        if _weight is None:
+            self._weight.process_group = ProcessGroup(tp_degree=self.world_size)
+            self._weight = ColoParameter.from_torch_tensor(torch.empty(self.num_embeddings,
+                                                                       self.embedding_dim_per_partition,
+                                                                       device='cpu',
+                                                                       dtype=dtype),
+                                                           requires_grad=True,
+                                                           spec=ShardSpec(dims=[-1], num_partitions=[self.world_size]))
+            self.init_parameters()
+        else:
+            assert isinstance(_weight, ColoParameter), "initialized weight must in type of ColoParameter"
+            _weight.process_group = ProcessGroup(tp_degree=self.world_size)
+            _weight.set_tensor_spec(ShardSpec(dims=[-1], num_partitions=[self.world_size]),
+                                    ComputeSpec(ComputePattern.TP1D))
+            self._weight = _weight
+
+    @property
+    def weight(self):
+        return self.cache_weight_mgr.cpu_weight
+
+    def named_parameters(self, prefix: str = '', recurse: bool = True) -> Iterator[Tuple[str, Parameter]]:
+        yield 'weight', self.cache_weight_mgr.cuda_cached_weight
+
+    def parameters(self, recurse: bool = True) -> Iterator[Parameter]:
+        yield self.cache_weight_mgr.cuda_cached_weight
+
+    @torch.no_grad()
+    def init_parameters(self):
+        self._weight.data.uniform_(-1 / self.num_embeddings, 1 / self.num_embeddings)
+        if self.padding_idx is not None:
+            self._weight[self.padding_idx].fill_(0)
+
+    def preprocess(self,
+                   cuda_row_num: int,
+                   ids_freq_mapping: Optional[List[int]] = None,
+                   warmup_ratio: float = 0.7,
+                   buffer_size: int = 50_000):
+        self.cache_weight_mgr = CachedParamMgr(self._weight, cuda_row_num, buffer_size=buffer_size)
+        self.cache_weight_mgr.reorder(ids_freq_mapping, warmup_ratio)
+
+    def forward(self, indices, offsets=None, per_sample_weights=None, shape_hook=None, scatter_dim=0, gather_dim=-1):
+        with torch.no_grad():
+            reorder_ids = self.cache_weight_mgr.prepare_ids(indices)
+
+        output_shard = F.embedding_bag(reorder_ids, self.cache_weight_mgr.cuda_cached_weight, offsets, self.max_norm,
+                                       self.norm_type, self.scale_grad_by_freq, self.mode, self.sparse,
+                                       per_sample_weights, self.include_last_offset, self.padding_idx)
+
+        if shape_hook is not None:
+            output_shard = shape_hook(output_shard)
```

Comment:
the shape hook might introduce some tensor view&squeeze operations.
I suppose we should be aware of that when using ColoParameter.

Reply:
OK, can you update it in another PR? I copy this line from yours code.

Reply:
OK